src/move/emojicoin_arena/sources/emojicoin_arena.move

```diff
@@ -0,0 +1,703 @@
+    const E_NEW_DURATION_TOO_SHORT: u64 = 1;
+    /// New melee lock-in period is too long.
+    const E_NEW_LOCK_IN_PERIOD_TOO_LONG: u64 = 2;
+    /// New melee match percentage is too high.
+    const E_NEW_MATCH_PERCENTAGE_TOO_HIGH: u64 = 3;
+    /// User's melee escrow has nonzero emojicoin 0 balance.
+    const E_ENTER_COIN_BALANCE_0: u64 = 4;
+    /// User's melee escrow has nonzero emojicoin 1 balance.
+    const E_ENTER_COIN_BALANCE_1: u64 = 5;
+    /// Elected to lock in but unable to match.
+    const E_UNABLE_TO_LOCK_IN: u64 = 6;
+    /// Provided escrow coin type is invalid.
+    const E_INVALID_ESCROW_COIN_TYPE: u64 = 7;
+
+    /// Resource account address seed for the registry.
+    const REGISTRY_SEED: vector<u8> = b"Arena registry";
+
+    const U64_MAX: u64 = 0xffffffffffffffff;
+    const MAX_PERCENTAGE: u64 = 100;
+    const MIN_MELEE_DURATION: u64 = 1 * 3_600_000_000;
+
+    /// Flat integrator fee.
+    const INTEGRATOR_FEE_RATE_BPS: u8 = 100;
+    const INTEGRATOR_FEE_RATE_BPS_DUAL_ROUTE: u8 = 50;
+
+    // Default parameters for new melees.
+    const DEFAULT_DURATION: u64 = 36 * 3_600_000_000;
+    const DEFAULT_AVAILABLE_REWARDS: u64 = 1000 * 100_000_000;
+    const DEFAULT_MAX_MATCH_PERCENTAGE: u64 = 50;
+    const DEFAULT_MAX_MATCH_AMOUNT: u64 = 5 * 100_000_000;
+
+    struct Melee has store {
+        /// 1-indexed for conformity with emojicoin market ID indexing.
+        melee_id: u64,
+        /// Metadata for market with lower market ID comes first.
+        market_metadatas: vector<MarketMetadata>,
+        /// In microseconds.
+        start_time: u64,
+        /// How long melee lasts after start time.
+        duration: u64,
+        /// Amount of rewards available for distribution in octas, conditional on vault balance.
+        available_rewards: u64,
+        /// Max percentage of user's input amount to match in octas, when locking in.
+        max_match_percentage: u64,
+        /// Maximum amount of APT to match in octas, when locking in.
+        max_match_amount: u64,
+        /// All entrants who have entered the melee.
+        all_entrants: SmartTable<address, Nil>,
+        /// Active entrants in the melee.
+        active_entrants: SmartTable<address, Nil>,
+        /// Entrants who have exited the melee.
+        exited_entrants: SmartTable<address, Nil>,
+        /// Entrants who have locked in.
+        locked_in_entrants: SmartTable<address, Nil>,
+        /// Number of melee-specific swaps.
+        n_melee_swaps: Aggregator<u64>,
+        /// Volume of melee-specific swaps in octas.
+        melee_swaps_volume: Aggregator<u128>,
+        /// Amount of emojicoin 0 locked in all melee escrows for the melee.
+        emojicoin_0_locked: Aggregator<u64>,
+        /// Amount of emojicoin 1 locked in all melee escrows for the melee.
+        emojicoin_1_locked: Aggregator<u64>
+    }
+
+    struct MeleeEscrow<phantom Coin0, phantom LP0, phantom Coin1, phantom LP1> has key {
```

Comment:
```suggestion

    invariant<Coin0, LP0, Coin1, LP1> forall account_address: address where exists<
        MeleeEscrow<Coin0, LP0, Coin1, LP1>>(account_address):
        coin::value<Coin0>(
            global<MeleeEscrow<Coin0, LP0, Coin1, LP1>>(account_address).emojicoin_0
        ) == 0
            || coin::value<Coin1>(
                global<MeleeEscrow<Coin0, LP0, Coin1, LP1>>(account_address).emojicoin_1
            ) == 0;

    struct MeleeEscrow<phantom Coin0, phantom LP0, phantom Coin1, phantom LP1> has key {
```

@wrwg I am trying to use this global invariant to enforce that all coins are in 1 of the two types, but it is erroring with:

```
{
  "Error": "Move Prover failed: [internal] boogie exited with compilation errors:\n/Users/alex/repos/emojicoin-dot-fun/src/move/emojicoin_arena/boogie.bpl(32395,288): Error: cannot refer to a global variable in this context: #1_info\n/Users/alex/repos/emojicoin-dot-fun/src/move/emojicoin_arena/boogie.bpl(32395,300): Error: cannot refer to a global variable in this context: #1_info\n/Users/alex/repos/emojicoin-dot-fun/src/move/emojicoin_arena/boogie.bpl(32395,312): Error: cannot refer to a global variable in this context: #1_info\n/Users/alex/repos/emojicoin-dot-fun/src/move/emojicoin_arena/boogie.bpl(32395,462): Error: cannot refer to a global variable in this context: #1_info\n/Users/alex/repos/emojicoin-dot-fun/src/move/emojicoin_arena/boogie.bpl(32395,474): Error: cannot refer to a global variable in this context: #1_info\n/Users/alex/repos/emojicoin-dot-fun/src/move/emojicoin_arena/boogie.bpl(32395,486): Error: cannot refer to a global variable in this context: #1_info\n/Users/alex/repos/emojicoin-dot-fun/src/move/emojicoin_arena/boogie.bpl(32400,288): Error: cannot refer to a global variable in this context: #2_info\n/Users/alex/repos/emojicoin-dot-fun/src/move/emojicoin_arena/boogie.bpl(32400,300): Error: cannot refer to a global variable in this context: #2_info\n/Users/alex/repos/emojicoin-dot-fun/src/move/emojicoin_arena/boogie.bpl(32400,312): Error: cannot refer to a global variable in this context: #2_info\n/Users/alex/repos/emojicoin-dot-fun/src/move/emojicoin_arena/boogie.bpl(32400,462): Error: cannot refer to a global variable in this context: #2_info\n/Users/alex/repos/emojicoin-dot-fun/src/move/emojicoin_arena/boogie.bpl(32400,474): Error: cannot refer to a global variable in this context: #2_info\n/Users/alex/repos/emojicoin-dot-fun/src/move/emojicoin_arena/boogie.bpl(32400,486): Error: cannot refer to a global variable in this context: #2_info\n/Users/alex/repos/emojicoin-dot-fun/src/move/emojicoin_arena/boogie.bpl(32405,288): Error: cannot refer to a global variable in this context: #3_info\n/Users/alex/repos/emojicoin-dot-fun/src/move/emojicoin_arena/boogie.bpl(32405,300): Error: cannot refer to a global variable in this context: #3_info\n/Users/alex/repos/emojicoin-dot-fun/src/move/emojicoin_arena/boogie.bpl(32405,312): Error: cannot refer to a global variable in this context: #3_info\n/Users/alex/repos/emojicoin-dot-fun/src/move/emojicoin_arena/boogie.bpl(32405,462): Error: cannot refer to a global variable in this context: #3_info\n/Users/alex/repos/emojicoin-dot-fun/src/move/emojicoin_arena/boogie.bpl(32405,474): Error: cannot refer to a global variable in this context: #3_info\n/Users/alex/repos/emojicoin-dot-fun/src/move/emojicoin_arena/boogie.bpl(32405,486): Error: cannot refer to a global variable in this context: #3_info\n18 name resolution errors detected in /Users/alex/repos/emojicoin-dot-fun/src/move/emojicoin_arena/boogie.bpl\n"
}
```

Even if I comment out the invariant. E.g. it looks like the prover fails even when I have no specifications

Reply:
Tracked at https://github.com/aptos-labs/aptos-core/issues/15605